// ngxs-reset-plugin/esm2015/lib/symbols.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/symbols.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { isDevMode } from '@angular/core';
import { getStoreMetadata } from '@ngxs/store';
/**
 * Action to clear all state except given state(s)
 */
export class StateClear {
    /**
     * @param {...?} statesToKeep
     */
    constructor(...statesToKeep) {
        /** @type {?} */
        const reducer = createMetaDataListReducer(isDevMode());
        this.statesToKeep = statesToKeep.reduce(reducer, []);
    }
}
StateClear.type = '@@CLEAR_STATE';
if (false) {
    /** @type {?} */
    StateClear.type;
    /** @type {?} */
    StateClear.prototype.statesToKeep;
}
/**
 * Action to reset given state(s) to defaults
 */
export class StateReset {
    /**
     * @param {...?} statesToReset
     */
    constructor(...statesToReset) {
        /** @type {?} */
        const reducer = createMetaDataListReducer(isDevMode());
        this.statesToReset = statesToReset.reduce(reducer, []);
    }
}
StateReset.type = '@@RESET_STATE';
if (false) {
    /** @type {?} */
    StateReset.type;
    /** @type {?} */
    StateReset.prototype.statesToReset;
}
/**
 * Action to reset all states expect given state(s) to defaults
 */
export class StateResetAll {
    /**
     * @param {...?} statesToKeep
     */
    constructor(...statesToKeep) {
        /** @type {?} */
        const reducer = createMetaDataListReducer(isDevMode());
        this.statesToKeep = statesToKeep.reduce(reducer, []);
    }
}
StateResetAll.type = '@@RESET_STATE_ALL';
if (false) {
    /** @type {?} */
    StateResetAll.type;
    /** @type {?} */
    StateResetAll.prototype.statesToKeep;
}
/**
 * Action to overwrite state(s) with given value(s)
 */
export class StateOverwrite {
    /**
     * @param {...?} overwriteConfigs
     */
    constructor(...overwriteConfigs) {
        /** @type {?} */
        const reducer = createMetaTupleReducer(isDevMode());
        const [states, values] = overwriteConfigs.reduce(reducer, [
            [],
            [],
        ]);
        this.statesToOverwrite = states;
        this.values = values;
    }
}
StateOverwrite.type = '@@OVERWRITE_STATE';
if (false) {
    /** @type {?} */
    StateOverwrite.type;
    /** @type {?} */
    StateOverwrite.prototype.statesToOverwrite;
    /** @type {?} */
    StateOverwrite.prototype.values;
}
/**
 * @param {?} state
 * @param {?} devMode
 * @return {?}
 */
export function getMetaData(state, devMode) {
    /** @type {?} */
    const meta = (/** @type {?} */ (new Object(getStoreMetadata((/** @type {?} */ (state))))));
    /** @type {?} */
    const isNgxsMeta = meta.name && 'defaults' in meta;
    // Reusability Hack: devMode is number on purpose
    if (!isNgxsMeta && devMode === -2) {
        console.warn(`Reset Plugin Warning: ${meta.name} is not a state class.`);
        return null;
    }
    return meta;
}
/**
 * @param {?} devMode
 * @return {?}
 */
function createMetaDataListReducer(devMode) {
    return (/**
     * @param {?} acc
     * @param {?} state
     * @return {?}
     */
    (acc, state) => {
        // tslint:disable-next-line:no-bitwise
        /** @type {?} */
        const meta = getMetaData(state, ~devMode);
        return meta ? acc.concat(meta) : acc;
    });
}
/**
 * @param {?} devMode
 * @return {?}
 */
function createMetaTupleReducer(devMode) {
    return (/**
     * @param {?} acc
     * @param {?} __1
     * @return {?}
     */
    (acc, [state, value]) => {
        // tslint:disable-next-line:no-bitwise
        /** @type {?} */
        const meta = getMetaData(state, ~devMode);
        return meta ? [acc[0].concat(meta), acc[1].concat(value)] : acc;
    });
}
