// @ngxs/logger-plugin/__ivy_ngcc__/fesm2015/ngxs-logger-plugin.js
import { InjectionToken, Injectable, Inject, Injector, NgModule } from '@angular/core';
import { getActionTypeFromInstance, Store, NGXS_PLUGINS } from '@ngxs/store';
import { tap, catchError } from 'rxjs/operators';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
import * as ɵngcc0 from '@angular/core';
const repeat = (/**
 * @param {?} str
 * @param {?} times
 * @return {?}
 */
(str, times) => new Array(times + 1).join(str));
/** @type {?} */
const pad = (/**
 * @param {?} num
 * @param {?} maxLength
 * @return {?}
 */
(num, maxLength) => repeat('0', maxLength - num.toString().length) + num);
/**
 * @param {?} time
 * @return {?}
 */
function formatTime(time) {
    return (pad(time.getHours(), 2) +
        `:` +
        pad(time.getMinutes(), 2) +
        `:` +
        pad(time.getSeconds(), 2) +
        `.` +
        pad(time.getMilliseconds(), 3));
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ActionLogger {
    /**
     * @param {?} action
     * @param {?} store
     * @param {?} logWriter
     */
    constructor(action, store, logWriter) {
        this.action = action;
        this.store = store;
        this.logWriter = logWriter;
    }
    /**
     * @param {?} state
     * @return {?}
     */
    dispatched(state) {
        /** @type {?} */
        const actionName = getActionTypeFromInstance(this.action);
        /** @type {?} */
        const formattedTime = formatTime(new Date());
        /** @type {?} */
        const message = `action ${actionName} @ ${formattedTime}`;
        this.logWriter.startGroup(message);
        // print payload only if at least one property is supplied
        if (this._hasPayload(this.action)) {
            this.logWriter.logGrey('payload', Object.assign({}, this.action));
        }
        this.logWriter.logGrey('prev state', state);
    }
    /**
     * @param {?} nextState
     * @return {?}
     */
    completed(nextState) {
        this.logWriter.logGreen('next state', nextState);
        this.logWriter.endGroup();
    }
    /**
     * @param {?} error
     * @return {?}
     */
    errored(error) {
        this.logWriter.logRedish('next state after error', this.store.snapshot());
        this.logWriter.logRedish('error', error);
        this.logWriter.endGroup();
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    _hasPayload(event) {
        /** @type {?} */
        const nonEmptyProperties = this._getNonEmptyProperties(event);
        return nonEmptyProperties.length > 0;
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    _getNonEmptyProperties(event) {
        /** @type {?} */
        const keys = Object.keys(event);
        /** @type {?} */
        const values = keys.map((/**
         * @param {?} key
         * @return {?}
         */
        key => event[key]));
        return values.filter((/**
         * @param {?} value
         * @return {?}
         */
        value => value !== undefined));
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    ActionLogger.prototype.action;
    /**
     * @type {?}
     * @private
     */
    ActionLogger.prototype.store;
    /**
     * @type {?}
     * @private
     */
    ActionLogger.prototype.logWriter;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class LogWriter {
    /**
     * @param {?} options
     */
    constructor(options) {
        this.options = options;
        this.options = this.options || (/** @type {?} */ ({}));
        this.logger = options.logger || console;
    }
    /**
     * @param {?} message
     * @return {?}
     */
    startGroup(message) {
        /** @type {?} */
        const startGroupFn = this.options.collapsed
            ? this.logger.groupCollapsed
            : this.logger.group;
        try {
            startGroupFn.call(this.logger, message);
        }
        catch (e) {
            console.log(message);
        }
    }
    /**
     * @return {?}
     */
    endGroup() {
        try {
            this.logger.groupEnd();
        }
        catch (e) {
            this.logger.log('—— log end ——');
        }
    }
    /**
     * @param {?} title
     * @param {?} payload
     * @return {?}
     */
    logGrey(title, payload) {
        /** @type {?} */
        const greyStyle = 'color: #9E9E9E; font-weight: bold';
        this.log(title, greyStyle, payload);
    }
    /**
     * @param {?} title
     * @param {?} payload
     * @return {?}
     */
    logGreen(title, payload) {
        /** @type {?} */
        const greenStyle = 'color: #4CAF50; font-weight: bold';
        this.log(title, greenStyle, payload);
    }
    /**
     * @param {?} title
     * @param {?} payload
     * @return {?}
     */
    logRedish(title, payload) {
        /** @type {?} */
        const redishStyle = 'color: #FD8182; font-weight: bold';
        this.log(title, redishStyle, payload);
    }
    /**
     * @param {?} title
     * @param {?} color
     * @param {?} payload
     * @return {?}
     */
    log(title, color, payload) {
        if (this.isIE()) {
            this.logger.log(title, payload);
        }
        else {
            this.logger.log('%c ' + title, color, payload);
        }
    }
    /**
     * @return {?}
     */
    isIE() {
        /** @type {?} */
        const ua = typeof window !== 'undefined' && window.navigator.userAgent
            ? window.navigator.userAgent
            : '';
        /** @type {?} */
        let msIE = false;
        /** @type {?} */
        const oldIE = ua.indexOf('MSIE ');
        /** @type {?} */
        const newIE = ua.indexOf('Trident/');
        if (oldIE > -1 || newIE > -1) {
            msIE = true;
        }
        return msIE;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    LogWriter.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    LogWriter.prototype.options;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
function NgxsLoggerPluginOptions() { }
if (false) {
    /**
     * Auto expand logged actions
     * @type {?|undefined}
     */
    NgxsLoggerPluginOptions.prototype.collapsed;
    /**
     * Provide alternate console.log implementation
     * @type {?|undefined}
     */
    NgxsLoggerPluginOptions.prototype.logger;
    /**
     * Disable the logger. Useful for prod mode.
     * @type {?|undefined}
     */
    NgxsLoggerPluginOptions.prototype.disabled;
    /**
     * Predicate for actions to be the logged. Takes action and state snapshot as parameters
     * @type {?|undefined}
     */
    NgxsLoggerPluginOptions.prototype.filter;
}
/** @type {?} */
const NGXS_LOGGER_PLUGIN_OPTIONS = new InjectionToken('NGXS_LOGGER_PLUGIN_OPTIONS');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NgxsLoggerPlugin {
    /**
     * @param {?} _options
     * @param {?} _injector
     */
    constructor(_options, _injector) {
        this._options = _options;
        this._injector = _injector;
    }
    /**
     * @param {?} state
     * @param {?} event
     * @param {?} next
     * @return {?}
     */
    handle(state, event, next) {
        if (this._options.disabled || !(/** @type {?} */ (this._options.filter))(event, state)) {
            return next(state, event);
        }
        this._logWriter = this._logWriter || new LogWriter(this._options);
        // Retrieve lazily to avoid cyclic dependency exception
        this._store = this._store || this._injector.get(Store);
        /** @type {?} */
        const actionLogger = new ActionLogger(event, this._store, this._logWriter);
        actionLogger.dispatched(state);
        return next(state, event).pipe(tap((/**
         * @param {?} nextState
         * @return {?}
         */
        nextState => {
            actionLogger.completed(nextState);
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        error => {
            actionLogger.errored(error);
            throw error;
        })));
    }
}
NgxsLoggerPlugin.ɵfac = function NgxsLoggerPlugin_Factory(t) { return new (t || NgxsLoggerPlugin)(ɵngcc0.ɵɵinject(NGXS_LOGGER_PLUGIN_OPTIONS), ɵngcc0.ɵɵinject(ɵngcc0.Injector)); };
NgxsLoggerPlugin.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: NgxsLoggerPlugin, factory: NgxsLoggerPlugin.ɵfac });
/** @nocollapse */
NgxsLoggerPlugin.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [NGXS_LOGGER_PLUGIN_OPTIONS,] }] },
    { type: Injector }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgxsLoggerPlugin, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [NGXS_LOGGER_PLUGIN_OPTIONS]
            }] }, { type: ɵngcc0.Injector }]; }, null); })();
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgxsLoggerPlugin.prototype._store;
    /**
     * @type {?}
     * @private
     */
    NgxsLoggerPlugin.prototype._logWriter;
    /**
     * @type {?}
     * @private
     */
    NgxsLoggerPlugin.prototype._options;
    /**
     * @type {?}
     * @private
     */
    NgxsLoggerPlugin.prototype._injector;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const USER_OPTIONS = new InjectionToken('LOGGER_USER_OPTIONS');
/**
 * @param {?} options
 * @return {?}
 */
function loggerOptionsFactory(options) {
    /** @type {?} */
    const defaultLoggerOptions = {
        logger: console,
        collapsed: false,
        disabled: false,
        filter: (/**
         * @return {?}
         */
        () => true)
    };
    return Object.assign({}, defaultLoggerOptions, options);
}
class NgxsLoggerPluginModule {
    /**
     * @param {?=} options
     * @return {?}
     */
    static forRoot(options) {
        return {
            ngModule: NgxsLoggerPluginModule,
            providers: [
                {
                    provide: NGXS_PLUGINS,
                    useClass: NgxsLoggerPlugin,
                    multi: true
                },
                {
                    provide: USER_OPTIONS,
                    useValue: options
                },
                {
                    provide: NGXS_LOGGER_PLUGIN_OPTIONS,
                    useFactory: loggerOptionsFactory,
                    deps: [USER_OPTIONS]
                }
            ]
        };
    }
}
NgxsLoggerPluginModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxsLoggerPluginModule });
NgxsLoggerPluginModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxsLoggerPluginModule_Factory(t) { return new (t || NgxsLoggerPluginModule)(); } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgxsLoggerPluginModule, [{
        type: NgModule
    }], null, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { NGXS_LOGGER_PLUGIN_OPTIONS, NgxsLoggerPlugin, NgxsLoggerPluginModule, USER_OPTIONS as ɵa, loggerOptionsFactory as ɵb };
