// @ngxs/logger-plugin/esm5/src/action-logger.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { getActionTypeFromInstance } from '@ngxs/store';
import { formatTime } from './internals';
var ActionLogger = /** @class */ (function () {
    function ActionLogger(action, store, logWriter) {
        this.action = action;
        this.store = store;
        this.logWriter = logWriter;
    }
    /**
     * @param {?} state
     * @return {?}
     */
    ActionLogger.prototype.dispatched = /**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        /** @type {?} */
        var actionName = getActionTypeFromInstance(this.action);
        /** @type {?} */
        var formattedTime = formatTime(new Date());
        /** @type {?} */
        var message = "action " + actionName + " @ " + formattedTime;
        this.logWriter.startGroup(message);
        // print payload only if at least one property is supplied
        if (this._hasPayload(this.action)) {
            this.logWriter.logGrey('payload', tslib_1.__assign({}, this.action));
        }
        this.logWriter.logGrey('prev state', state);
    };
    /**
     * @param {?} nextState
     * @return {?}
     */
    ActionLogger.prototype.completed = /**
     * @param {?} nextState
     * @return {?}
     */
    function (nextState) {
        this.logWriter.logGreen('next state', nextState);
        this.logWriter.endGroup();
    };
    /**
     * @param {?} error
     * @return {?}
     */
    ActionLogger.prototype.errored = /**
     * @param {?} error
     * @return {?}
     */
    function (error) {
        this.logWriter.logRedish('next state after error', this.store.snapshot());
        this.logWriter.logRedish('error', error);
        this.logWriter.endGroup();
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    ActionLogger.prototype._hasPayload = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var nonEmptyProperties = this._getNonEmptyProperties(event);
        return nonEmptyProperties.length > 0;
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    ActionLogger.prototype._getNonEmptyProperties = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var keys = Object.keys(event);
        /** @type {?} */
        var values = keys.map((/**
         * @param {?} key
         * @return {?}
         */
        function (key) { return event[key]; }));
        return values.filter((/**
         * @param {?} value
         * @return {?}
         */
        function (value) { return value !== undefined; }));
    };
    return ActionLogger;
}());
export { ActionLogger };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ActionLogger.prototype.action;
    /**
     * @type {?}
     * @private
     */
    ActionLogger.prototype.store;
    /**
     * @type {?}
     * @private
     */
    ActionLogger.prototype.logWriter;
}
