// @ngxs/logger-plugin/esm5/src/log-writer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var LogWriter = /** @class */ (function () {
    function LogWriter(options) {
        this.options = options;
        this.options = this.options || (/** @type {?} */ ({}));
        this.logger = options.logger || console;
    }
    /**
     * @param {?} message
     * @return {?}
     */
    LogWriter.prototype.startGroup = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
        /** @type {?} */
        var startGroupFn = this.options.collapsed
            ? this.logger.groupCollapsed
            : this.logger.group;
        try {
            startGroupFn.call(this.logger, message);
        }
        catch (e) {
            console.log(message);
        }
    };
    /**
     * @return {?}
     */
    LogWriter.prototype.endGroup = /**
     * @return {?}
     */
    function () {
        try {
            this.logger.groupEnd();
        }
        catch (e) {
            this.logger.log('—— log end ——');
        }
    };
    /**
     * @param {?} title
     * @param {?} payload
     * @return {?}
     */
    LogWriter.prototype.logGrey = /**
     * @param {?} title
     * @param {?} payload
     * @return {?}
     */
    function (title, payload) {
        /** @type {?} */
        var greyStyle = 'color: #9E9E9E; font-weight: bold';
        this.log(title, greyStyle, payload);
    };
    /**
     * @param {?} title
     * @param {?} payload
     * @return {?}
     */
    LogWriter.prototype.logGreen = /**
     * @param {?} title
     * @param {?} payload
     * @return {?}
     */
    function (title, payload) {
        /** @type {?} */
        var greenStyle = 'color: #4CAF50; font-weight: bold';
        this.log(title, greenStyle, payload);
    };
    /**
     * @param {?} title
     * @param {?} payload
     * @return {?}
     */
    LogWriter.prototype.logRedish = /**
     * @param {?} title
     * @param {?} payload
     * @return {?}
     */
    function (title, payload) {
        /** @type {?} */
        var redishStyle = 'color: #FD8182; font-weight: bold';
        this.log(title, redishStyle, payload);
    };
    /**
     * @param {?} title
     * @param {?} color
     * @param {?} payload
     * @return {?}
     */
    LogWriter.prototype.log = /**
     * @param {?} title
     * @param {?} color
     * @param {?} payload
     * @return {?}
     */
    function (title, color, payload) {
        if (this.isIE()) {
            this.logger.log(title, payload);
        }
        else {
            this.logger.log('%c ' + title, color, payload);
        }
    };
    /**
     * @return {?}
     */
    LogWriter.prototype.isIE = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var ua = typeof window !== 'undefined' && window.navigator.userAgent
            ? window.navigator.userAgent
            : '';
        /** @type {?} */
        var msIE = false;
        /** @type {?} */
        var oldIE = ua.indexOf('MSIE ');
        /** @type {?} */
        var newIE = ua.indexOf('Trident/');
        if (oldIE > -1 || newIE > -1) {
            msIE = true;
        }
        return msIE;
    };
    return LogWriter;
}());
export { LogWriter };
if (false) {
    /**
     * @type {?}
     * @private
     */
    LogWriter.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    LogWriter.prototype.options;
}
