// @ngxs/store/esm2015/src/public_to_deprecate.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { getSelectorMetadata as getSelectorMetadataInternal, getStoreMetadata as getStoreMetadataInternal, ensureStoreMetadata as ensureStoreMetadataInternal, ensureSelectorMetadata as ensureSelectorMetadataInternal } from './internal/internals';
/**
 * @record
 */
function MetaDataModel() { }
if (false) {
    /** @type {?} */
    MetaDataModel.prototype.name;
    /** @type {?} */
    MetaDataModel.prototype.actions;
    /** @type {?} */
    MetaDataModel.prototype.defaults;
    /** @type {?} */
    MetaDataModel.prototype.path;
    /** @type {?|undefined} */
    MetaDataModel.prototype.children;
}
/**
 * @record
 */
function SelectorMetaDataModel() { }
if (false) {
    /** @type {?} */
    SelectorMetaDataModel.prototype.originalFn;
    /** @type {?} */
    SelectorMetaDataModel.prototype.containerClass;
    /** @type {?} */
    SelectorMetaDataModel.prototype.selectorName;
    /** @type {?} */
    SelectorMetaDataModel.prototype.getSelectorOptions;
}
/**
 * @param {?} target
 * @return {?}
 */
export function ensureStoreMetadata(target) {
    return ensureStoreMetadataInternal(target);
}
/**
 * @param {?} target
 * @return {?}
 */
export function getStoreMetadata(target) {
    return getStoreMetadataInternal(target);
}
/**
 * @param {?} target
 * @return {?}
 */
export function ensureSelectorMetadata(target) {
    return ensureSelectorMetadataInternal(target);
}
/**
 * @param {?} target
 * @return {?}
 */
export function getSelectorMetadata(target) {
    return getSelectorMetadataInternal(target);
}
