// @ngxs/logger-plugin/esm2015/src/action-logger.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { getActionTypeFromInstance } from '@ngxs/store';
import { formatTime } from './internals';
export class ActionLogger {
    /**
     * @param {?} action
     * @param {?} store
     * @param {?} logWriter
     */
    constructor(action, store, logWriter) {
        this.action = action;
        this.store = store;
        this.logWriter = logWriter;
    }
    /**
     * @param {?} state
     * @return {?}
     */
    dispatched(state) {
        /** @type {?} */
        const actionName = getActionTypeFromInstance(this.action);
        /** @type {?} */
        const formattedTime = formatTime(new Date());
        /** @type {?} */
        const message = `action ${actionName} @ ${formattedTime}`;
        this.logWriter.startGroup(message);
        // print payload only if at least one property is supplied
        if (this._hasPayload(this.action)) {
            this.logWriter.logGrey('payload', Object.assign({}, this.action));
        }
        this.logWriter.logGrey('prev state', state);
    }
    /**
     * @param {?} nextState
     * @return {?}
     */
    completed(nextState) {
        this.logWriter.logGreen('next state', nextState);
        this.logWriter.endGroup();
    }
    /**
     * @param {?} error
     * @return {?}
     */
    errored(error) {
        this.logWriter.logRedish('next state after error', this.store.snapshot());
        this.logWriter.logRedish('error', error);
        this.logWriter.endGroup();
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    _hasPayload(event) {
        /** @type {?} */
        const nonEmptyProperties = this._getNonEmptyProperties(event);
        return nonEmptyProperties.length > 0;
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    _getNonEmptyProperties(event) {
        /** @type {?} */
        const keys = Object.keys(event);
        /** @type {?} */
        const values = keys.map((/**
         * @param {?} key
         * @return {?}
         */
        key => event[key]));
        return values.filter((/**
         * @param {?} value
         * @return {?}
         */
        value => value !== undefined));
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    ActionLogger.prototype.action;
    /**
     * @type {?}
     * @private
     */
    ActionLogger.prototype.store;
    /**
     * @type {?}
     * @private
     */
    ActionLogger.prototype.logWriter;
}
